### PR TITLE
Add `can_setindex` methods

### PR DESCRIPTION
**Pull request by Tokazama**

### CODE REVIEW

**Review by chriselrod — APPROVED** [review]
Of course, some of these can `Base.setindex`, but `can_setindex` is already about `setindex!` instead.

File: src/ArrayInterface.jl [hunk 1]
```diff
@@ -187,6 +187,11 @@ Query whether a type can use `setindex!`.
 can_setindex(x) = can_setindex(typeof(x))
 can_setindex(::Type) = true
 can_setindex(::Type{<:AbstractRange}) = false
+can_setindex(::Type{<:AbstractDict}) = true
+can_setindex(::Type{<:Base.ImmutableDict}) = false
+can_setindex(@nospecialize T::Type{<:Tuple}) = false
+can_setindex(@nospecialize T::Type{<:NamedTuple}) = false
+can_setindex(::Type{<:Base.Pairs{<:Any,<:Any,P}}) where {P} = can_setindex(P)
```
Comment (chriselrod) [inline src/ArrayInterface.jl:194]: Not defined in Julia 1.6, apparently.

**Comment by Tokazama** [comment]
> Of course, some of these can `Base.setindex`, but `can_setindex` is already about `setindex!` instead.

Is your point that `can_setindex` could be confused as referring to `Base.setindex`?

File: src/ArrayInterface.jl (hunk 1 above)
Comment (Tokazama) [reply src/ArrayInterface.jl:194]: Just realized it's in `Iterators.Pairs`

**Comment by chriselrod** [comment]
> > Of course, some of these can `Base.setindex`, but `can_setindex` is already about `setindex!` instead.
> 
> Is your point that `can_setindex` could be confused as referring to `Base.setindex`?

Yeah. We already have `can_setindex` meaning this though. Was just making an observation.

**Comment by Tokazama** [comment]
> > > Of course, some of these can `Base.setindex`, but `can_setindex` is already about `setindex!` instead.
> > 
> > 
> > Is your point that `can_setindex` could be confused as referring to `Base.setindex`?
> 
> Yeah. We already have `can_setindex` meaning this though. Was just making an observation.

It's a good point and looking at the blame history I don't think it was discussed in much detail when implemented. I can make an issue for it just in case anyone feels strongly about this or wants to approach that issue in another PR.

**Comment by ChrisRackauckas** [comment]
And one of the most important cases for this would be CuArrays, which should get an overload that is dependent on the global.

**Comment by Tokazama** [comment]
I'm assuming that would have to be implemented with something like `CUDA.context`.

**Pull request merged**